cEP-0025.md

```diff
@@ -0,0 +1,169 @@
+[Asteroid](https://github.com/PyCQA/astroid) lies in its simplicity
+and provision of lots of re-usable analysis. Asteroid is more detailed and
+powerful, thus making it complicated. Plus, there is a much larger ecosystem
+of flake8 plugins.
+
+## Proposed Change
+
+Here is a brief overview of the architecture:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install coala-pyflakes`.
+2. The repository will contain a new package `pyflakes-plugin-bears` which
+   would house all bears which use pyflakes-enhanced-ast.
+3. The `pyflakes-plugin-bears` would contain a metabear `PyFlakesASTBear`
+   which would be used by all other plugin bears.
+4. The repository would even contain a new package `pyflakes-generic-plugins`
+   that would be independent of coala and could be run using flake8.
+
+## Management of the new `coala-pyflakes` repository
+
+The repository will use a similar mechanism of what is being
+used in `coala-bears` to test all its bear plugins.
+
+## Implementation of PyFlakesASTBear
+
+1. PyFlakesASTBear will be LocalBear returning result of the type
+   HiddenResult since these results are not meant for the users but
+   for the dependent bear. The result structure is described by
+   PyFlakesResult class.
+2. PyFlakesResult consists of a snapshot of the module at class,
+   function, module, generator ,and doctest level. Thus, providing
+   the developer multiple views of a file allowing him to retrieve
+   results as per his demand.
+3. Apart from scope information, the metabear even returns a list of
+   messages generated by pyflakes itself. These messages can be used
+   by the developer to add an additional filter layer that is
+   required to be resolved by the user before executing the plugin.
+   This is primarily done because pyflakes emits additional syntax
+   errors which are not caught by python interpretor and may affect
+   the working of a plugin. This check is optional and may or may not
+   be incorparated by the developer.
+4. Helper function `get_node_location` provides with node location
+   in source file.
+5. Helper function `get_nodes` provides with all the nodes of a
+   particular type present in a scope.
+
+Here is a prototype for the implementations of PyFlakesASTBear:
+
+```python
+# Imports not mentioned to maintain brevity
+
+class PyFlakesResult(HiddenResult):
+
+    def __init__(self, origin, deadScopes, pyflakes_messages):
+
+        Result.__init__(self, origin, message='')
+
+        self.class_scope = self.get_scopes(ClassScope, deadScopes)
+        self.function_scope = self.get_scopes(FunctionScope, deadScopes)
+        self.module_scope = self.get_scopes(ModuleScope, deadScopes)
+        self.generator_scope = self.get_scopes(GeneratorScope, deadScopes)
+        self.doctest_scope = self.get_scopes(DoctestScope, deadScopes)
+        self.pyflakes_messages = pyflakes_messages
+
+    def get_scopes(self, scope_level, scopes):
```

Comment:
scope_level -> scope_type.

many of these scopes can be nested in various ways, such as functions in a doctest.

We'll probably want to add some helper which makes it easier for plugins to skip anything in a doctest, as they are usually more complicated, and need different logic.

And this is why we probably need to support a visitor pattern, as plugins think about scopes from outer to the inner, and want to exclude large branches that dont interest them.  But the visitor pattern can be thought about later after we've built some plugins and find the pain points.

Actually this is something that is worth mentioning in more detail above.  I am pretty sure Python ast doesnt provide doctests, and pretty sure that pycodestyle also doesnt have any reasonable way to create plugins for doctest.  The pyflakes ability to provide doctest is a serious feature for projects like coala who use them extensively.  Currently errors in them are caught by pyflakes, but the style of them is mostly ignored.

Maybe this is another good example for the cEP, as you could easily show a plugin which is probably very difficult in any other framework.